File: src/main/java/arimitsu/sf/cql/v3/Frame.java
package arimitsu.sf.cql.v3;

import java.nio.ByteBuffer;

/**
 * Created by sxend on 14/06/04.
 */
public class Frame {
    public final Header header;
    public final int length;
    public final byte[] body;

    public Frame(ByteBuffer byteBuffer) {
        byte version = byteBuffer.get();
        byte flags = byteBuffer.get();
        short streamId = byteBuffer.getShort();
        byte opcode = byteBuffer.get();
        int length = byteBuffer.getInt();
        byte[] body = new byte[length];
        byteBuffer.get(body);
        this.header = new Header(Version.valueOf(version), flags, streamId, Opcode.valueOf(opcode));
        this.length = length;
        this.body = body;
    }

    public Frame(Header header, byte[] body) {
        this.header = header;
        this.body = body;
        this.length = body != null ? body.length : 0;
    }

    public static class Builder {
        public Header.Builder headerBuilder = new Header.Builder();
        public byte[] body;

        public Builder version(Version v) {
            this.headerBuilder = headerBuilder.version(v);
            return this;
        }

        public Builder flags(byte f) {
            this.headerBuilder = headerBuilder.flags(f);
            return this;
        }

        public Builder streamId(short s) {
            this.headerBuilder = headerBuilder.streamId(s);
            return this;
        }

        public Builder opcode(Opcode o) {
            this.headerBuilder = headerBuilder.opcode(o);
            return this;
        }

        public Builder body(byte[] b) {
            this.body = b;
            return this;
        }

        public Frame build() {
            return new Frame(this.headerBuilder.build(), this.body);
        }

    }

    public ByteBuffer toByteBuffer() {
        ByteBuffer byteBuffer = ByteBuffer.allocate(9 + this.length);
        byte[] headerBytes = new byte[9];
        headerBytes[0] = this.header.version.value;
        headerBytes[1] = this.header.flags;
        headerBytes[2] = (byte) (this.header.streamId >>> 8);
        headerBytes[3] = (byte) (0xff & this.header.streamId);
        headerBytes[4] = this.header.opcode.value;
        headerBytes[5] = (byte) (this.length >>> 24);
        headerBytes[6] = (byte) (0xff & (this.length >>> 16));
        headerBytes[7] = (byte) (0xff & (this.length >>> 8));
        headerBytes[8] = (byte) (0xff & this.length);
        byteBuffer.put(headerBytes);
        if (this.body != null) {
            byteBuffer.put(this.body);
        }
        return byteBuffer;
    }


}

File: src/main/scala/arimitsu/sf/cql/v3/Frame.scala
package arimitsu.sf.cql.v3

import java.nio.ByteBuffer

/**
 * Created by sxend on 14/06/04.
 */
object Frame {
  def apply(byteBuffer: ByteBuffer): Frame = {
    val version = byteBuffer.get()
    val flags = byteBuffer.get()
    val streamId = byteBuffer.getShort
    val opcode = byteBuffer.get()
    val header = new Header(Version.valueOf(version), flags, streamId, Opcode.valueOf(opcode))
    val length = byteBuffer.getInt
    Frame(header, length ,Option(byteBuffer))
  }

  def apply(header: Header,length:Int, body: Option[ByteBuffer]): Frame = {
    new Frame(header, length, body)
  }
}

class Frame(val header: Header, val length: Int, val body: Option[ByteBuffer]) {
  def toByteBuffer: ByteBuffer = {
    val byteBuffer = ByteBuffer.allocate(9 + this.length)
    val headerBytes = new Array[Byte](9)
    headerBytes.update(0, header.version.value)
    headerBytes.update(1, header.flags)
    headerBytes.update(2, (this.header.streamId >>> 8).toByte)
    headerBytes.update(3, (0xff & this.header.streamId).toByte)
    headerBytes.update(4, header.opcode.value)
    headerBytes.update(5, (this.length >>> 24).toByte)
    headerBytes.update(6, (0xff & (this.length >>> 16)).toByte)
    headerBytes.update(7, (0xff & (this.length >>> 8)).toByte)
    headerBytes.update(8, (0xff & this.length).toByte)
    byteBuffer.put(headerBytes)
    body match{
      case Some(buf) =>
        val arr = new Array[Byte](length)
        buf.get(arr)
        byteBuffer.put(arr)
      case None =>
    }
    byteBuffer.flip()
    byteBuffer
  }

}